LanguageFeatures/Patterns/null_assert_A03_t01.dart

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion nullAssertPattern ::= primaryPattern '!'
+///
+/// A null-assert pattern is similar to a null-check pattern in that it permits
+/// non-null values to flow through. But a null-assert throws if the matched
+/// value is null. It lets you forcibly assert that you know a value shouldn't
+/// be null, much like the corresponding ! null-assert expression.
+///
+/// @description Check that it is a static warning if matched value of a
+/// null-asssert pattern cannot ne null
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns
+
+String test1(int? x) {
+  switch (x) {
+    case final int? v1! || final int? v1!:
+//                                      ^
+// [analyzer] STATIC_WARNING.UNNECESSARY_NULL_ASSERT_PATTERN
+// [cfe] The null-assert pattern will have no effect because the matched type isn't nullable.
+      return "match-1";
+    case int? v2 || int? v2!:
+//                         ^
+// [analyzer] STATIC_WARNING.UNNECESSARY_NULL_ASSERT_PATTERN
+// [cfe] The null-assert pattern will have no effect because the matched type isn't nullable.
+      return "match-2";
+    case int v3!:
+//             ^
+// [analyzer] STATIC_WARNING.UNNECESSARY_NULL_ASSERT_PATTERN
+// [cfe] The null-assert pattern will have no effect because the matched type isn't nullable.
+      return "match-3";
+    default:
+      return "no match";
+  }
+}
+
+String test2(int? x) {
+  if (x case final int? v1! || final int? v1!)
+//                                          ^
+// [analyzer] STATIC_WARNING.UNNECESSARY_NULL_ASSERT_PATTERN
+// [cfe] The null-assert pattern will have no effect because the matched type isn't nullable.
+    return "match-1";
+  if (x case int? v2 || int? v2!)
+//                             ^
+// [analyzer] STATIC_WARNING.UNNECESSARY_NULL_ASSERT_PATTERN
+// [cfe] The null-assert pattern will have no effect because the matched type isn't nullable.
```

Comment:
I wonder if this is because the inferred type of `x` for the right hand operand of `||` is `Never`? In any case, it makes sense that it can't be null, but it would also make sense to report that the right hand operand of `||` is dead code. ...

Reply:
This was moved to a new test library, IIUC.